crazy_rl/multi_agent/jax/catch/catch.py

```diff
@@ -0,0 +1,300 @@
+            target_location=jnp.array([self._target_location]),
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def auto_reset(self, **state):
+        """Returns the State reinitialized if needed, else the actual State."""
+        done = jnp.any(state["truncations"]) + jnp.any(state["terminations"])
+
+        state = State(
+            agents_locations=done * self._init_flying_pos + (1 - done) * state["agents_locations"],
+            timestep=(1 - done) * state["timestep"],
+            observations=state["observations"],
+            rewards=(1 - done) * state["rewards"],
+            terminations=(1 - done) * state["terminations"],
+            truncations=(1 - done) * state["truncations"],
+            target_location=done * self._target_location + (1 - done) * state["target_location"],
+        )
+        state = self._compute_obs(state)
+        return state
+
+    @partial(jit, static_argnums=(0,))
+    def state_to_dict(self, state):
+        """Translates the State into a dict."""
+        return {
+            "agents_locations": state.agents_locations,
+            "timestep": state.timestep,
+            "observations": state.observations,
+            "rewards": state.rewards,
+            "terminations": state.terminations,
+            "truncations": state.truncations,
+            "target_location": state.target_location,
+        }
+
+    @partial(jit, static_argnums=(0,))
+    def step_vmap(self, action, key, **state_val):
+        """Calls step with a State and is called by vmap without State object."""
+        return self.step(State(**state_val), action, key)
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def state(self, state):
+        """Returns a global observation (concatenation of all the agent locations and target locations)."""
+        return jnp.append(state.agents_locations.flatten(), state.target_location)
+
+
+if __name__ == "__main__":
+    from jax.lib import xla_bridge
+
+    jax.config.update("jax_platform_name", "gpu")
+
+    print(xla_bridge.get_backend().platform)
+
+    parallel_env = Catch(
+        num_drones=5,
+        init_flying_pos=jnp.array([[0.0, 0.0, 1.0], [2.0, 1.0, 1.0], [0.0, 1.0, 1.0], [2.0, 2.0, 1.0], [1.0, 0.0, 1.0]]),
+        init_target_location=jnp.array([1.0, 1.0, 2.5]),
+        target_speed=0.1,
+    )
+
+    n = 1000  # number of states in parallel
+    seed = 5  # test value
+    key = random.PRNGKey(seed)
+
+    @jit
+    def body(i, states_key):
+        """Body of the fori_loop of play.
+
+        Args:
+            i: number of the iteration.
+            states_key: a tuple containing states and key.
+        """
+        actions = random.uniform(states_key[1], (n, parallel_env.num_drones, 3), minval=-1, maxval=1)
+
+        key, *subkeys = random.split(states_key[1], n + 1)
+
+        states = vmap(parallel_env.step_vmap)(actions, jnp.stack(subkeys), **parallel_env.state_to_dict(states_key[0]))
```

Comment:
I think we can store the partially applied functions into their proper variables to avoid transforming the function everytime:

`vmapped_step = vmap(parallel_env.step_vmap)`

then call 
`states =  vmapped_step(actions, jnp.stack(subkeys), **parallel_env.state_to_dict(states_key[0]))`